### ISSUE TITLE
the usage of clusters in the training 

### ISSUE THREAD

**Issue by hawkjk** [peiyunh/tiny#5]
Hello Peiyun:

I found that the clusters in your demo as blow:

  -86.3382 -113.4441   86.3382  113.4441    0.5000
  -48.7500  -65.2500   48.7500   65.2500    0.5000
  -33.2500  -43.7500   33.2500   43.7500    0.5000
  -25.7500  -33.7500   25.7500   33.7500    0.5000
  -40.5000  -54.5000   40.5000   54.5000    1.0000
  -34.5000  -43.5000   34.5000   43.5000    1.0000
  -28.5000  -38.0000   28.5000   38.0000    1.0000
  -25.6589  -31.3221   25.6589   31.3221    1.0000
  -21.6137  -27.5977   21.6137   27.5977    1.0000
  -20.0000  -22.5000   20.0000   22.5000    1.0000
  -17.5000  -25.5000   17.5000   25.5000    1.0000
  -16.3280  -20.8856   16.3280   20.8856    1.0000
  -29.4755  -34.4803   29.4755   34.4803    2.0000
  -25.4202  -37.1061   25.4202   37.1061    2.0000
  -24.2118  -30.2146   24.2118   30.2146    2.0000
  -22.0129  -24.7059   22.0129   24.7059    2.0000
  -19.3142  -28.0202   19.3142   28.0202    2.0000
  -17.9677  -22.7849   17.9677   22.7849    2.0000
  -15.6123  -19.5907   15.6123   19.5907    2.0000
  -13.1842  -18.3421   13.1842   18.3421    2.0000
  -13.5028  -15.0792   13.5028   15.0792    2.0000
  -11.1091  -16.4909   11.1091   16.4909    2.0000
  -10.9769  -14.0478   10.9769   14.0478    2.0000
   -9.9722  -12.4105    9.9722   12.4105    2.0000
   -9.6613  -10.6161    9.6613   10.6161    2.0000

I just want to know should I put all the 25 clusters into the training for all image pyramid(including 0.5X, 1X,2X) or should I only put the relative clusters base on the scale factor in the last column to the relative scaled image in training? I mean [0,4] for 0.5X images training, [5,12] for 1X images training and [13:24] for 2X image training?
Thanks.

**Comment by peiyunh** [comment]
Good question. I recommend training all 25 templates on every scale. My experience is that training templates only on corresponding scales performs slightly worse.

**Comment by hawkjk** [comment]
Thanks peiyun. I just want to reproduce the result of your paper base on the py caffe. But now it seems that the result has something wrong. I think I have achieved  most of ideas of your paper except more batch size and ohem. But the mAP in WiderFace validation set is only 6%. Too many false positives and bad performance for easy face. Could you give me some suggestions about that?  Hope to see your training code as soon as possible. Thank you!

**Comment by hawkjk** [comment]
Hello peiyun, I found that if I reuse the parameter in resnet between ohem layer and hard mining layer subsequently , the AP value drop down about 10 percent t. But all the ohem paper suggest to reuse parameter . 
Another problem is should I put the use_global_stats to be true or false of bn layer in training? It seems that how to implement the ohem in bn layer in resnet is very different with vgg16.

**Comment by peiyunh** [comment]
We did not implement the online hard negative mining in the same way as the OHEM paper did. Please refer to "Online hard mining and balanced sampling" in the appendix of [our paper](https://arxiv.org/pdf/1612.04402.pdf) for our implementation.

**Comment by hawkjk** [comment]
Hello peiyun, thanks for your clarification. I have another question is how to " adding context" in resnet network?  Firstly , should we use concat layer or Eltwise layer to adding context? Secondly , when I try to adding conv2, conv3 and conv5 , I found that the value of conv5's feature map in resnet is much larger than conv2 and conv3. If I adding these layers together, it seems that the featuremaps of  conv2 and conv3 can not take any effect on the final result. Should I scale the value to match conv5? Thanks!

**Comment by peiyunh** [comment]
Here is a visualization of our network: http://ethereon.github.io/netscope/#/gist/8a0d5ef37da9dc4cd611d178404b3641. Hopefully it will address your questions.
